### PR TITLE
[HDFS] Experimental Kerberos Support

### PR DESCRIPTION
**Pull request by mohitsoni**
- Adds experimental Kerberos Support via `SECURE_MODE` flag.
- Introduces `TASK_NAME` env var.
- Adds support for per task JRE via `TASK_JRE_URI`. This is important in Kerberos configuration so that JRE with pre-installed (inside `jre/lib/security`) JCE jars and `krb5.conf` can used.
- Enables better configurable logging via `HADOOP_OPTS` and `HADOOP_ROOT_LOGGER`.
- Experimental `KEYTABS_URI` is used for downloading keytabs on each node. This WILL BE replaced in future once better solution is available.
- Makes service user configurable.

Doc: https://docs.google.com/document/d/1lqlEIs98j1VsAyoEYnhYoaNmYylcoaBAwHpD29yKjU4

### CODE REVIEW

File: frameworks/hdfs/universe/config.json [hunk 1]
```diff
@@ -60,6 +60,30 @@
 					"type": "string",
 					"description": "The placement strategy for HDFS nodes. This should always be NODE.",
 					"default": "NODE"
+				},
+				"secure_mode": {
+					"id": "http://hdfs/dcos/mesosphere.com/service/secure_mode",
+					"type": "boolean",
+					"description": "Enable HDFS Secure Mode",
```
Comment (mgummelt) [inline frameworks/hdfs/universe/config.json]: Can you elaborate on what this means?

File: frameworks/hdfs/universe/config.json [hunk 2]
```diff
@@ -60,6 +60,30 @@
...
+					"default": false
+				},
+				"keytabs_uri": {
+					"id": "http://hdfs/dcos/mesosphere.com/service/keytabs_uri",
+					"type": "string",
+					"description": "The URI from which the Keytabs are downloaded.",
+					"default": "http://10.0.1.69:8788/keytabs.tar.gz"
```
Comment (mgummelt) [inline frameworks/hdfs/universe/config.json]: why is this the default?

File: frameworks/hdfs/universe/config.json [hunk 3]
```diff
@@ -60,6 +60,30 @@
...
+				},
+				"task_jre_uri": {
+					"id": "http://hdfs/dcos/mesosphere.com/service/task_jre_uri",
+					"type": "string",
+					"description": "The URI from which the JRE for all tasks is downloaded",
+					"default": "https://s3-us-west-2.amazonaws.com/infinity-artifacts/jre-8u91-linux-x64.tar.gz"
```
Comment (mgummelt) [inline frameworks/hdfs/universe/config.json]: Use the user-friendly s3 url (http://infinity-artifacts.s3.amazonaws.com/...)

File: frameworks/hdfs/universe/config.json (hunk 2 above)
Comment (mgummelt) [inline frameworks/hdfs/universe/config.json]: What is this a tarball of?  Please document

File: frameworks/hdfs/universe/config.json [hunk 4]
```diff
@@ -60,6 +60,30 @@
...
+				},
+				"user": {
+					"id": "http://hdfs/dcos/mesosphere.com/service/user",
+					"type": "string",
+					"description": "The user as which service and tasks are run.",
```
Comment (mgummelt) [inline frameworks/hdfs/universe/config.json]: This is odd wording.  Let's not be afraid of preposition-at-the-end-of-sentence: "The user the service and tasks are run as."

File: frameworks/hdfs/universe/config.json [hunk 5]
```diff
@@ -60,6 +60,30 @@
...
+				"secure_mode": {
+					"id": "http://hdfs/dcos/mesosphere.com/service/secure_mode",
+					"type": "boolean",
+					"description": "Enable HDFS Secure Mode",
+					"default": false
+				},
+				"keytabs_uri": {
+					"id": "http://hdfs/dcos/mesosphere.com/service/keytabs_uri",
+					"type": "string",
+					"description": "The URI from which the Keytabs are downloaded.",
```
Comment (mgummelt) [inline frameworks/hdfs/universe/config.json]: Document that this is only relevant if secure_mode is enabled.

File: frameworks/hdfs/universe/resource.json [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {
   "assets": {
     "uris": {
-      "jre-tar-gz": "https://downloads.mesosphere.com/kafka/assets/jre-8u91-linux-x64.tar.gz",
+      "jre-tar-gz": "https://s3-us-west-2.amazonaws.com/infinity-artifacts/jre-8u91-linux-x64.tar.gz",
```
Comment (mgummelt) [inline frameworks/hdfs/universe/resource.json]: We should use a production bucket, not infinity-artifacts.

Also, let's use the user-friendly S3 URLs I referenced above.

File: frameworks/hdfs/universe/config.json [hunk 7]
```diff
@@ -64,25 +64,25 @@
...
-					"default": "http://10.0.1.69:8788/keytabs.tar.gz"
+					"description": "The URI from which Keytabs tar.gz can be downloaded. Only relevant when secure_mode is enabled.",
+					"default": ""
 				},
 				"task_jre_uri": {
 					"id": "http://hdfs/dcos/mesosphere.com/service/task_jre_uri",
 					"type": "string",
 					"description": "The URI from which the JRE for all tasks is downloaded",
-					"default": "https://s3-us-west-2.amazonaws.com/infinity-artifacts/jre-8u91-linux-x64.tar.gz"
+					"default": "http://infinity-artifacts.s3.amazonaws.com/jre-8u91-linux-x64.tar.gz"
```
Comment (mgummelt) [inline frameworks/hdfs/universe/config.json]: we should serve everything over https, right?

File: frameworks/hdfs/universe/config.json [hunk 8]
```diff
@@ -64,25 +64,25 @@
 				"secure_mode": {
 					"id": "http://hdfs/dcos/mesosphere.com/service/secure_mode",
 					"type": "boolean",
-					"description": "Enable HDFS Secure Mode",
+					"description": "Enable HDFS Secure Mode, which allows configuration of Kerberos.",
```
Comment (mgummelt) [inline frameworks/hdfs/universe/config.json]: It doesn't "allow configuration", it actually does the configuration.  "This enables Kerberos authentication, as well as network encryption for data and RPC calls."

File: frameworks/hdfs/universe/config.json [hunk 9]
```diff
@@ -60,6 +60,30 @@
...
+					"id": "http://hdfs/dcos/mesosphere.com/service/keytabs_uri",
+					"type": "string",
+					"description": "The URI from which Keytabs tar.gz can be downloaded. Only relevant when secure_mode is enabled.",
+					"default": ""
+				},
+				"task_jre_uri": {
+					"id": "http://hdfs/dcos/mesosphere.com/service/task_jre_uri",
+					"type": "string",
+					"description": "The URI from which the JRE for all tasks is downloaded",
+					"default": "http://infinity-artifacts.s3.amazonaws.com/jre-8u91-linux-x64.tar.gz"
```
Comment (mgummelt) [inline frameworks/hdfs/universe/config.json]: Should not use infinity-artifacts in production

File: frameworks/hdfs/universe/config.json [hunk 10]
```diff
@@ -60,6 +60,30 @@
 					"type": "string",
 					"description": "The placement strategy for HDFS nodes. This should always be NODE.",
 					"default": "NODE"
+				},
+				"secure_mode": {
```
Comment (mgummelt) [inline frameworks/hdfs/universe/config.json]: This is ambiguous.  Let's just call this "kerberos".

**Comment by gabrielhartmann** [comment]
@mohitsoni:  Can we finish this off or close the PR, please?

**Pull request closed without merging**